client/my-sites/earn/components/add-edit-coupon-modal/test/index.jsx

```diff
@@ -0,0 +1,353 @@
+		__.mockImplementation( mockTranslate );
+		getconnectedAccountDefaultCurrencyForSiteId.mockImplementation(
+			mockGetConnectedAccountDefaultCurrencyForSiteId
+		);
+		getconnectedAccountMinimumCurrencyForSiteId.mockImplementation(
+			mockGetConnectedAccountMinimumCurrencyForSiteId
+		);
+
+		modalRoot = document.createElement( 'div' );
+		modalRoot.setAttribute( 'id', 'wpcom' );
+		document.body.appendChild( modalRoot );
+		Modal.setAppElement( modalRoot );
+	} );
+
+	afterEach( () => {
+		unmountComponentAtNode( modalRoot );
+		document.body.removeChild( modalRoot );
+		[ ...document.getElementsByClassName( 'ReactModalPortal' ) ].map( ( el ) => {
+			document.body.removeChild( el );
+		} );
+		modalRoot = null;
+		global.Date = RealDate;
+	} );
+
+	test( 'should render empty form when no coupon is provided', () => {
+		mockDate( '2023-06-23T14:23:44z' );
+		render(
+			<WrappedRecurringPaymentsCouponAddEditModal closeDialog={ closeDialog } coupon={ {} } />
+		);
+		const codeTextbox = getCodeTextbox();
+		const amountOption = getAmountOption();
+		const percentageOption = getPercentageOption();
+		const amountInput = getAmountInput();
+		const limitDurationToggle = getLimitDurationToggle();
+		const limitDurationSelection = getLimitDurationSelection();
+		const limitCouponsToggle = getLimitCouponsToggle();
+		const limitCouponsTextInput = getLimitCouponsTextInput();
+
+		expect( codeTextbox.value ).toEqual( '' );
+		expect( amountOption.selected ).toBe( false );
+		expect( percentageOption.selected ).toBe( true );
+		expect( amountInput ).name = 'discount_percentage';
+		expect( amountInput ).type = 'text';
+		expect( amountInput ).value = '';
+		expect( limitDurationToggle ).not.toBeChecked();
+		expect( limitDurationSelection ).toBeDisabled();
+		expect( limitCouponsToggle ).not.toBeChecked();
+		expect( limitCouponsTextInput ).toBeDisabled();
+	} );
+
+	test( 'should generate a new coupon code when the random button is clicked', () => {
+		mockDate( '2023-06-23T14:23:44z' );
+		render(
+			<WrappedRecurringPaymentsCouponAddEditModal closeDialog={ closeDialog } coupon={ {} } />
+		);
+		const randomButton = getRandomButton();
+		const codeTextbox = getCodeTextbox();
+
+		expect( codeTextbox.value ).toEqual( '' );
+		waitFor( () => randomButton.click() );
+		waitFor( () => {
+			const generatedCode = codeTextbox.value;
+			expect( generatedCode ).not.toEqual( '' );
+			expect( generatedCode.length ).toBeGreaterThan( 3 );
+		} );
+	} );
+
+	test( 'should show discount amount input when discount type is amount', async () => {
+		mockDate( '2023-06-23T14:23:44z' );
+		render(
+			<WrappedRecurringPaymentsCouponAddEditModal closeDialog={ closeDialog } coupon={ {} } />
+		);
+		const discountTypeSelect = getDiscountTypeSelect();
+		const amountOption = getAmountOption();
+		const percentageOption = getPercentageOption();
+		const amountInput = getAmountInput();
+
+		expect( amountOption.selected ).toBe( false );
```

Comment:
We should use `toHaveAttribute()` or even better, if we're working with `select`, we should be working with asserting the value of the wrapping `select` field, otherwise we're testing implementation details which isn't recommended.

Reply:
Agreed. it helps to think from the perspective of what the consumer/user expects to interact with/see.